src/parser.js

```diff
@@ -125,188 +126,237 @@ export default class Parser {
+      // Swagger has a paths object to loop through that describes resources
       _.each(_.omit(swagger.paths, isExtension), (pathValue, href) => {
         this.handleSwaggerPath(pathValue, href);
       });
 
-      done(null, parseResult);
+      done(null, this.result);
     });
   }
 
   // == Internal properties & functions ==
 
-  handleSwaggerPath(pathValue, href) {
-    const {
-      Asset, Copy, Category, DataStructure, HrefVariables, HttpHeaders,
-      Member: MemberElement, Object: ObjectElement, Resource, Transition,
-    } = this.minim.elements;
-    const resource = new Resource();
+  // Lazy-loaded input AST is made available when we need it. If it can't be
+  // loaded, then an annotation is generated with more information about why.
+  get ast() {
+    if (this._ast !== undefined) {
+      return this._ast;
+    }
 
-    this.path.push('paths');
-    this.path.push(href);
+    if (_.isString(this.source)) {
+      try {
+        this._ast = new Ast(this.source);
+      } catch (err) {
+        this._ast = null;
+        this.createAnnotation(ANNOTATIONS.AST_UNAVAILABLE, null,
+          'Input AST could not be composed, so source maps will not be available');
+      }
+    } else {
+      this._ast = null;
+      this.createAnnotation(ANNOTATIONS.AST_UNAVAILABLE, null,
+        'Source maps are only available with string input');
+    }
 
-    if (this.generateSourceMap) {
-      this.makeSourceMap(resource, this.path);
+    return this._ast;
+  }
+
+  // This method lets you set the current parsing path and synchronously run
+  // a function (e.g. to create an element). If the function returns one or
+  // more elements then they will get a source map if one was requested. Once
+  // finished, the path is restored to its original value.
+  withPath(...args) {
+    let i;
+
+    for (i = 0; i < args.length - 1; i++) {
+      this.path.push(args[i]);
     }
 
-    // Provide users with a way to add a title to a resource in Swagger
-    if (pathValue['x-summary']) {
-      resource.title = pathValue['x-summary'];
+    let elements = args[args.length - 1].bind(this)(this.path);
+
+    if (elements && this.generateSourceMap) {
+      // You can return either an element or an array of elements.
+      if (!_.isArray(elements)) {
+        elements = [elements];
+      }
+      for (i of elements) {
+        this.createSourceMap(i, this.path);
+      }
     }
 
-    // Provide users a way to add a description to a resource in Swagger
-    if (pathValue['x-description']) {
-      const resourceDescription = new Copy(pathValue['x-description']);
-      resource.push(resourceDescription);
+    for (i = 0; i < args.length - 1; i++) {
+      this.path.pop();
     }
+  }
 
-    if (this.useResourceGroups()) {
-      const groupName = pathValue['x-group-name'];
+  // This is like `withPath` above, but slices the path before calling by
+  // using the first argument as a length (starting at index 0).
+  withSlicedPath(...args) {
```

Comment:
Heh, the behavior here is different though - it slices *before* so that you can remove items from the current path. It's only used in one place right now. After running, the `this.path` is the same instance as when it started, so at completion the behavior is consistent with the method above. E.g.

```js
let foo = this.path;

this.withPath('bar', () => {});
assert.equal(foo, this.path); // True

this.withSlicedPath(1, 'baz', () => {});
assert.equal(foo, this.path); // True
```